### PR TITLE
Apple Silicon User Experience

### PR DESCRIPTION
**Pull request by sdmaclea**
Moving dotnet/runtime#48628 here.

This is a verbatim copy of the initial draft for Apple Silicon from dotnet/runtime#48628. I will address the feedback in subsequent commits.

I have dropped the Windows UX draft version at least for now...  If there is something to say, I can create a separate PR.

I have put the docs in a user-experience folder...

### CODE REVIEW

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+#### dotnet-sos and the SOS plugin
+
+The `dotnet-sos` tool installs the `SOS` diagnostic plugin into `lldb`. Since `lldb` is a universal binary, it expects its plugins to also be universal binaries. The plan would be to make the Apple Silcon SOS be a universal binary to allow SOS to work on `x86_64` and `arm64` processes.
+
+#### Apple Store requires universal binaries
+
+Publishing a .NET runtime to the Apple store is not a common customer scenario. However it is worth noting that the Apple store is requiring universal binaries.
+
+This in currently an unsupported scenario.
```
Comment (akoeplinger) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:105]: Xamarin.Mac/MacCatalyst apps do support publishing to the App Store and there are plans to use CoreCLR for it which might impact this. /cc @marek-safar

Comment (sdmaclea) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:105]: @davidwrighton Is the crossgen2 file format robust enough to allow crossgen to two different architectures in the same file for supporting macOS universal binaries.  I assumed it wasn't.

Comment (jkotas) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:105]: Right, neither IL nor R2R file formats are designed as fat binaries. We would either need to have all non-native files SxS on disk, or we would need to work to evolve these file formats.

Comment (sdmaclea) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:105]: We could also evolve deps.json and the assembly load algorithm to support arch specific managed files.  (Unless it is already supported)

Comment (jkotas) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:105]: Yep,  deps.json supports the SxS option to large degree today. I am not sure whether core runtime files like CoreLib are handled correctly, but fixing that should be bug-fix level change.

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+The impact to the shared `dotnet` host is discussed at length above. `dotnet` will be a universal binary.
+
+#### dotnet-sos and the SOS plugin
+
+The `dotnet-sos` tool installs the `SOS` diagnostic plugin into `lldb`. Since `lldb` is a universal binary, it expects its plugins to also be universal binaries. The plan would be to make the Apple Silcon SOS be a universal binary to allow SOS to work on `x86_64` and `arm64` processes.
+
+#### Apple Store requires universal binaries
+
+Publishing a .NET runtime to the Apple store is not a common customer scenario. However it is worth noting that the Apple store is requiring universal binaries.
```
Comment (marek-safar) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:103]: I'd remove or rewrite this sense as it's incorrect. Today all Xamarin developers do exactly this.

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+This is largely TBD. The addition of a message during publish is the most likely design.
+
+## Universal Binaries
+
+As part of the architectural transition, Apple has switched the entire BigSur OS to use universal binaries. This means every file in the OS has both the `arm64` and the `x86_64` code on disk.
+
+Given both architectures, universal binaries are relatively easy to create. The general syntax would be `lipo <binaryArchA> <binaryArchB> -create -output <binaryUniversal>`
+
+### No planned universal binary .NET runtime
```
Comment (marek-safar) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:74]: What does .NET runtime refer to? I guess you meant the shared runtime host

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+At least for .NET 6, general universal binary support is not planned.
```
Comment (marek-safar) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:76]: I'd suggest rewording that. The universal binary support works already or is planned for Apple apps (bundles).

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+General support would require:
```
Comment (marek-safar) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:78]: Instead of general support, the list looks like console app specific needs

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+The `dotnet-sos` tool installs the `SOS` diagnostic plugin into `lldb`. Since `lldb` is a universal binary, it expects its plugins to also be universal binaries. The plan would be to make the Apple Silcon SOS be a universal binary to allow SOS to work on `x86_64` and `arm64` processes.
+
+#### Apple Store requires universal binaries
+
+Publishing a .NET runtime to the Apple store is not a common customer scenario. However it is worth noting that the Apple store is requiring universal binaries.
+
+This in currently an unsupported scenario.
+
+It may be possible for a developer to generate a universal binary .NET 6 app, if each architecture is published as a single file app. Each architecture could be merged into a universal binary. This may just work, but it may also invalidate assumptions made by the single file app.
```
Comment (marek-safar) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:107]: I don't follow the distinction, did you have some kind of coreclr specific single-exe console in mind here?

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+- The user or the IDE can explicitly select the architecture of the target process using the `arch` command.
+
+The extra search cost could be minimized by passing the result of the previous search when restarting the process. Adding an `--fx-version` option would accomplish this at the cost of changing the users specified argument order.  Using a new environment variable would avoid this unexpected side-effect.
+
+## Apple Silicon requires Signed Binaries
+
+Apple Silicon with Big Sur adds a new security policy which requires signed binaries by default. Unsigned binaries are simply killed without explanation. The user can expect an inscrutable message like `zsh killed: helloWorld` This directly affects our users who expect published binaries to just work.
+
+The developer would need to explicitly sign after publishing for Apple Silicon.  The sigining command is relatively simple `codesign -s <signature> <app-path>`.
+
```
Comment (marek-safar) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:63]: I believe macos workload will handle that automatically.

@rolfbjarne for confirmation

Comment (rolfbjarne) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:63]: Yes, code signing is handled automatically for Xamarin workloads.

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+The `dotnet-sos` tool installs the `SOS` diagnostic plugin into `lldb`. Since `lldb` is a universal binary, it expects its plugins to also be universal binaries. With the single architecture design, when `SOS` is installed, `lldb` will complain about not being able to load the plugin when the architecture mismatches.
+
+#### Universal Binary Experiment
+
+Making the Apple Silcon SOS be a universal binary to allowed SOS to work on `x86_64` and `arm64` processes.
+
+## User Experience Designs
+
+### Mixed Architecture Side By Side Installs
```
Comment (sdmaclea) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:173]: I would like to drive to consensus around this section.  At least from here to the primary design alternative described in Open Questions below.

As I wrote the Open Questions, it was clear the second alternative may be easier and perhaps more preferable.  This alternative may make it easier to support publishing universal binaries.

**Comment by sdmaclea** [comment]
I have substantially rewritten this doc.  I would appreciate more feedback.

File: accepted/2021/User-Experience/Apple-Silicon-UX.md (hunk 8 above)
Comment (jkotas) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:173]: How are other runtimes/SDKs with global installs planning to address this problem?

It may be useful to look at Java, golang, python, rust, ... to see whether there is a trend.

Comment (sdmaclea) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:173]: - I couldn't find an Oracle JDK for Apple Silicon.
- Zulu OpenJDK was arch specific.
- golang was arch specific.
- Python a universal binary. I think Apple provided the one I looked at
- Rust looks like it is not ready. They are still protoyping.

Comment (sdmaclea) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:173]: Our muxer architecture, readytorun story, and legacy mixed architecture make it more complicated.

Comment (jkotas) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:173]: From this list, it sounds like arch specific is the norm and we won't be an outlier if we stick with the simpler arch-specific as well.

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+
+#### Signing
+
+- dotnet publish for Xamarin workloads handles signing automatically.
+- dotnet publish for CoreClr console apps do not handle signing automatically. The developer would need to explicitly sign after publishing for Apple Silicon.  The signing command is relatively simple `codesign -s <signature> --entitlements <app-entitlements-path> <app-path>`.
+- There is no message to the customer that the published app needs to be signed
+
+#### Dotnet Publish as Universal Binaries
+
+- This scenario is also supported for Xamarin iOS workloads. This is a slightly different scenario as Xamarin iOS apps are native binaries (not JIT binaries)
```
Comment (rolfbjarne) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:160]: Xamarin.Mac is what runs on Apple Silicon (and it's jitted):

```suggestion
- This scenario is also supported for Xamarin Mac workloads.
```

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+
+- The experiment represents a viable design alternative. It did not support side-by-side install of multiple architectures of the same runtime version.
+
+- The implementation was incomplete. It did not automate the build process or deal with installer issues
+
+### Dotnet Publish Experience Inconsistent
+
+#### Signing
+
+- dotnet publish for Xamarin workloads handles signing automatically.
```
Comment (marek-safar) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:154]:
```suggestion
- dotnet publish for UI based workloads handles signing automatically.
```

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+- dotnet publish for CoreClr console apps do not handle signing automatically. The developer would need to explicitly sign after publishing for Apple Silicon.  The signing command is relatively simple `codesign -s <signature> --entitlements <app-entitlements-path> <app-path>`.
```
Comment (marek-safar) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:155]:
```suggestion
- dotnet publish for console apps do not handle signing automatically. The developer would need to explicitly sign after publishing for Apple
```

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+- There is no message to the customer that the published app needs to be signed
+
+#### Dotnet Publish as Universal Binaries
+
+- This scenario is also supported for Xamarin iOS workloads. This is a slightly different scenario as Xamarin iOS apps are native binaries (not JIT binaries)
+- CoreCLR runtime apps have no mechanism to publish universal binaries.
```
Comment (marek-safar) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:161]: Is that the correct message if we considering coreclr JIT to be backend for Xamarin.Mac ?

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+
+Proposal is currently _Unplanned/Uncommitted._
+
+### Publishing Universal Binary Apps
+
+No planned support for publishing CoreCLR universal binaries in .NET 6.
+
+This represent the current CoreCLR runtime team plan.
+
+Pursuing a universal CoreCLR runtime would affect this design and might allow us to at least specify a manual method to produce a universal app.
```
Comment (marek-safar) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:248]: I'd be better to fold this into https://github.com/dotnet/designs/pull/182/files#diff-f9955dc1a401114b9383fdfecab5a55da35312f33af1af21d4d46a0246fa7e16R158 section. It also needs clarification of what is exactly not supported.

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+
+While the customer can install the macOS Apple Silicon .NET preview build by itself, install this beside the macOS x64 builds is problematic.
+
+We have several possible representative scenarios. Two typical ones are described:
+
+- .NET 3.1 x64 with .NET 6 Apple Silicon. As side effect of not back-porting older releases to the new architecture is that customer will need to deal with multi-architecture side by side installs at least until .NET 3.1 goes out of support. However, this mixed forced architecture problem is transient in nature.
+
+- .NET 6 x64 with .NET 6 Apple Silicon. This scenario is not absolutely critical, but may be required by some customers.
+  - It enables moving legacy Apps with x64 dependent apps to .NET 6.
+  - It enables testing macOS x64 apps under the Rosetta 2 emulator.
```
Comment (vitek-karas) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:87]: Running apps via `dotnet app` if both `x64` and `arm64` is supported by the `dotnet` - this is currently problematic because there's nothing in the `app.runtimeconfig.json` which would hint the `dotnet`/`hostfxr` which architecture the app needs.
- We would either have to add this information somewhere (`.runtimeconfig.json` is probably the most natural place for this)
- Rely on `app` (executable) instead
  - Currently this is disabled by default for MacOS RID targets (if I remember correctly)  - this fact is not discussed anywhere in this doc either
  - Requires signing - as mentioned above

The universal muxer should default to one architecture if it doesn't know better (probably arm64)

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+  - Create a single distribution for both macOS architectures.
+  - Make the runtime native binaries to be universal binaries
+  - Modify `hostpolicy` to probe for trusted platform assemblies based on the current RID. This is designed to architecture specific ready-to-run managed assemblies. This RID specific path would be probed before probing the common any platform assembly path.
+  - Modify the runtime probing to be consistent with the new `hostpolicy` probing.
+- Modify the build system to build the Universal Binaries
+- Modify installers
+  - Modify .NET 5 installers to allow installation without breaking universal `dotnet`.
+  - Modify .NET 6 installer to overwrite the x64 `dotnet`
+
+#### Open Questions
```
Comment (vitek-karas) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:193]: My personal preference:
- Choose a different location for installation of `x64` .NET 6+ runtime - something like `dotnet_x64`. The existence of this is transient and in some not so distant future release will go away
- The primary install location for .NET 6+ will be `arm64` only (for the most part)
  - Keep the muxer/hostfxr as universal binaries in the main install location
  - Add some logic to also look for runtimes/SDKs in the `dotnet_x64` location - the universal muxer/hostfxr would use this to get the right runtime if asked to run `x64` app
- The primary location would be the only one registered (in config file and so on) and it would be the only one on "path".

File: accepted/2021/User-Experience/Apple-Silicon-UX.md (hunk 14 above)
Comment (jkotas) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:87]: > there's nothing in the app.runtimeconfig.json which would hint the dotnet/hostfxr which architecture the app needs.

This is intentional design. We talk about these as [cross-platform apps that can run anywhere](https://docs.microsoft.com/en-us/dotnet/core/deploying/#produce-a-cross-platform-binary).

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+  - Modify `hostfxr` to assume .NET 5 and earlier require x64 architecture.
+  - Modify `dotnet` to restart in the correct architecture on an architecture mismatch.
+- Make the CoreCLR runtime effectively a Universal Binaries
+  - Create a single distribution for both macOS architectures.
+  - Make the runtime native binaries to be universal binaries
+  - Modify `hostpolicy` to probe for trusted platform assemblies based on the current RID. This is designed to architecture specific ready-to-run managed assemblies. This RID specific path would be probed before probing the common any platform assembly path.
+  - Modify the runtime probing to be consistent with the new `hostpolicy` probing.
+- Modify the build system to build the Universal Binaries
+- Modify installers
+  - Modify .NET 5 installers to allow installation without breaking universal `dotnet`.
```
Comment (jkotas) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:190]:
```suggestion
  - Modify .NET 3.1 and .NET 5 installers to allow installation without breaking universal `dotnet`.
```
? .NET 5 goes out of support short after .NET 6 ships. .NET 3.1 is supported for much longer.

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+
+For macOS, universal are the default approach. Most binaries are multi-arch. For single arch binaries the kernel can detect and automatically switch the process to run in the correct architecture, either native or Rosetta 2 emulation of x64.
+
+A universal binary can be created simply by combining two or more architecture specific binaries into a universal binary using the `lipo` tool. The process is reversible too. A universal binary can be `thinned` to extract a single arch binaries.
+
+Given both architectures, universal binaries are relatively easy to create. The general syntax would be `lipo <binaryArchA> <binaryArchB> -create -output <binaryUniversal>`
+
+#### Apple Store requires universal binaries
+
+I have heard that publishing native Apple Silicon (or Apps supporting multiple architectures) apps to the Apple store will require them to be published as Universal binaries. I have not been able to find public documentation of this, so the details are not clear.
```
Comment (jkotas) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:31]: @rolfbjarne @marek-safar Is this requirement documented anywhere?

File: accepted/2021/User-Experience/Apple-Silicon-UX.md [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+
+When running on Apple Silicon unsigned binaries will get killed automatically by the kernel. A simple non-obvious message will be printed on the console (`zsh killed: helloWorld`).
+
+To allow a native shared library to be loaded, it will also need to be signed.
+
+When running the equivalent unsigned binary from a Rosetta 2 emulated process the process is not killed. When running a x64 process on Apple Silicon the process was not killed either.
+
+XCode 12 adhoc signs native binaries. These anonymous signature are sufficient to prevent these from being killed (on the machine that compiled them at least.)
+
+I haven't found the documentation exactly how and what changed. These behavior descriptions have been through experimentation and debugging.
```
Comment (jkotas) [inline accepted/2021/User-Experience/Apple-Silicon-UX.md:55]: Have you seen the work we have done on this? https://docs.microsoft.com/en-us/dotnet/core/install/macos-notarization-issues

> When running the equivalent unsigned binary from a Rosetta 2 emulated process the process is not killed.

I think that this is temporary to allow easier transition and it will be locked down.

File: accepted/2021/User-Experience/Apple-Silicon-UX.md (hunk 12 above)
Comment (jkotas) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:161]: Right, we would need to make this work as part of adding support for coreclr w/ Xamarin.Mac.

File: accepted/2021/User-Experience/Apple-Silicon-UX.md (hunk 17 above)
Comment (rolfbjarne) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:31]: I haven't heard anything about it, but it wouldn't surprise me if Apple made it required at some point.

Comment (marek-safar) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:31]: AFAIK it's the default setup for xcode projects but still a recommendation. I agree with Rolf that this can change in the future.

File: accepted/2021/User-Experience/Apple-Silicon-UX.md (hunk 6 above)
Comment (sdmaclea) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md]: Yes.  One of the coreclr publishing types is internally called single-file

File: accepted/2021/User-Experience/Apple-Silicon-UX.md (hunk 3 above)
Comment (sdmaclea) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md]: I was referring to the runtime distribution which would include the host, the coreclr, jit, vm ....

File: accepted/2021/User-Experience/Apple-Silicon-UX.md (hunk 18 above)
Comment (sdmaclea) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:55]: @odhanson Might be able to provide more context.

Comment (odhanson) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:55]: I will try gather more information both about the signing and the Apple store requirements and comment back here.

File: accepted/2021/User-Experience/Apple-Silicon-UX.md (hunk 14 above)
Comment (vitek-karas) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:87]: I know of the existence of "portable apps" - that's why the muxer will need a "default" architecture (for those apps). The point was about current state of things where we don't generate `apphost` for apps on MacOS and thus they're executed via `dotnet app.dll`. Those apps may not be portable, but currently there's no way to tell which RID they were built for (if any) and will break in bad ways if run on the wrong architecture (for example if they have native dependencies).

If we solve the signing problem and turn on `apphost` by default for MacOS, then this problem becomes much less important and we can probably continue to ignore it (just like we ignore it on Win x86/x64 today).

File: accepted/2021/User-Experience/Apple-Silicon-UX.md (hunk 18 above)
Comment (odhanson) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:55]: Here is a link:
https://developer.apple.com/documentation/macos-release-notes/macos-big-sur-11_0_1-universal-apps-release-notes

Code Signing
New Features
New in macOS 11 on Macs with Apple silicon, and starting in macOS Big Sur 11 beta 6, the operating system enforces that any executable must be signed before it’s allowed to run. There isn’t a specific identity requirement for this signature: a simple ad-hoc signature is sufficient. This new behavior doesn’t change the long-established policy that our users and developers can run arbitrary code on their Macs, and is designed to simplify the execution policies on Macs with Apple silicon and enable the system to better detect code modifications. This new policy doesn’t apply to translated x86 binaries running under Rosetta 2, nor does it apply to macOS 11 running on Intel-based platforms.

File: accepted/2021/User-Experience/Apple-Silicon-UX.md (hunk 17 above)
Comment (odhanson) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:31]: not yet formally required by Apple, but something we fully expect to be required later this year. As an example, Apple started mandating that all apps have 64-bit slices in their universal binaries about a year before they started telling people that 32-bit was ending. We know they will stop building Intel-based Macs within the next 18 months or so, so we expect them to start requiring Arm slices this summer.

File: accepted/2021/User-Experience/Apple-Silicon-UX.md (hunk 15 above)
Comment (AaronRobinsonMSFT) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:193]: After reading this document, the above statement by @vitek-karas would be my preference for a solution.

Comment (sdmaclea) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:193]: > Choose a different location for installation of x64 .NET 6+ runtime - something like dotnet_x64.

Are you proposing we backport this change to the 3.1/5.0 branches.

I was thinking we could only do this for the .NET 6+ branches.

Comment (sdmaclea) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:193]: I guess the advantage of @vitek-karas's approach would be that we just need to teach the installer to alter the install path for x64 on Apple Silicon.

That seems easier that fixing it overwriting the dotnet and hostfxr universal binaries.

That said, I am leanig to @vitek's approach too.

Comment (vitek-karas) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:193]: If we never install .NET 6+ x64 into global location (we simply don't produce the installers for this) then the `dotnet_x64` approach is not much different from keeping everything in `dotnet`. Obviously this is ignoring the potential problems with overwriting the `dotnet` (note that `hostfxr` doesn't have a problem since it's installed into a version specific directory, so the older installers will never overwrite the newer one).

The only true advantage of having `dotnet_x64` is that it would allow to install both .NET 6 x64 and .NET 6 arm64 side by side.

Comment (AaronRobinsonMSFT) [reply accepted/2021/User-Experience/Apple-Silicon-UX.md:193]: >The only true advantage of having dotnet_x64 is that it would allow to install both .NET 6 x64 and .NET 6 arm64 side by side.

It would also give us a pattern to follow - which is important here. Having a process that we can fallback on is important for our customers and ourselves when we rediscover this problem in 18 months.

**Comment by sdmaclea** [comment]
I believe @vitek-karas is going to drive this.  Closing.
